In [ ]:
# --- Бібліотеки для даної роботи ---
try:
    import numpy, pandas, matplotlib
    print("Бібліотеки вже встановлені. Пропускаємо інсталяцію.")
except ImportError:
    print("Встановлюємо бібліотеки...")
    %pip install numpy pandas matplotlib

Бібліотеки вже встановлені. Пропускаємо інсталяцію.


## Задача 2. Пошук схожих фільмів

**Умова:**

Рекомендаційні системи, як Netflix або YouTube, часто представляють користувачів та контент у вигляді векторів.

Є профіль користувача — це вектор інтересів до жанрів: `[Екшн, Комедія, Драма]`.

Також є три фільми $A, B, C$, кожен з яких має таку ж структуру характеристик.

Вектор профілю користувача: $\mathbf{u} = (8, 2, 5)$

Та три фільми:
* Фільм A (Action Movie): $\mathbf{v}_A = (9, 1, 2)$
* Фільм B (Comedy Movie): $\mathbf{v}_B = (1, 9, 8)$
* Фільм C (Drama Movie): $\mathbf{v}_C = (7, 2, 6)$

Ваша задача - знайти фільм, який найбільше підходить користувачу, порівнявши вектори їхніх характеристик.

> **Нагадуємо:** чим менший кут між цими векторами у багатовимірному просторі, тим більше схожі об'єкти. Мірою цієї схожості є **косинусна подібність**. Вона показує, наскільки вектори "дивляться" в один бік. Значення коливається від -1 (протилежні) до 1 (ідентичні).

Формула косинусної подібності:
$$\cos \theta = \frac{\mathbf{u} \cdot \mathbf{v}}{\|\mathbf{u}\| \cdot \|\mathbf{v}\|}$$
де $\mathbf{u} \cdot \mathbf{v}$ — скалярний добуток, а $\|\mathbf{u}\|, \|\mathbf{v}\|$ — довжина (норма) вектора.

**Аналітична частина розв'язку:**
1.  Обчисліть аналітично норму вектора користувача $\|\mathbf{u}\|$.
2.  Обчисліть схожість (similarity score) профілю користувача з кожним із трьох фільмів.
3.  Порівняйте отримані значення та визначте, який фільм має найвищий бал схожості.

**Програмна частина розв'язку:**
Використовуючи бібліотеку `numpy`, напишіть програму та виконайте наступні кроки:
1.  Створіть вектор профілю користувача та вектори фільмів.
2.  Реалізуйте функцію косинусної схожості за формулою.
3.  Обчисліть схожість для кожного фільму.
4.  Виведіть на екран коефіцієнти схожості для кожного фільму та назву фільму, вектор якого має найбільше значення косинусної подібності.
5.  Перевірте себе та порівняйте з аналітичними обчисленнями, щоб вони збігалися.

In [ ]:
import numpy as np
import pandas as pd
from IPython.display import display

user_preferences = np.array([8, 2, 5])

movies_database = {
    "Фільм A (Екшн)": np.array([9, 1, 2]),
    "Фільм B (Комедія)": np.array([1, 9, 8]),
    "Фільм C (Драма)  ": np.array([7, 2, 6])
}

def analyze_similarity(vector_a, vector_b):
    dot_product = np.dot(vector_a, vector_b)
    norm_a = np.linalg.norm(vector_a)
    norm_b = np.linalg.norm(vector_b)
    similarity_score = dot_product / (norm_a * norm_b)
    return dot_product, norm_b, similarity_score

results_list = []

print(f"Вектор користувача: {user_preferences}")
print(f"Довжина вектора (Norm): {np.linalg.norm(user_preferences):.4f}\n")

print("Технічний вивід:")
print(f"{'Назва фільму':<20} | {'Скалярний добуток':<18} | {'Норма вектора':<14} | {'Схожість':<10}")
print("-" * 70)

for title, movie_vector in movies_database.items():
    dot_val, norm_val, score = analyze_similarity(user_preferences, movie_vector)

    print(f"{title:<20} | {dot_val:<18.2f} | {norm_val:<14.4f} | {score:.4f}")

    results_list.append({
        "Назва фільму": title,
        "Вектор": str(movie_vector),
        "Схожість": score
    })

print("-" * 70)

print("Красивий Вивід:")
df_results = pd.DataFrame(results_list).sort_values(by="Схожість", ascending=False)
display(df_results.style.background_gradient(subset=["Схожість"], cmap="Greens"))

Вектор користувача: [8 2 5]
Довжина вектора (Norm): 9.6437

Технічний вивід:
Назва фільму         | Скалярний добуток  | Норма вектора  | Схожість  
----------------------------------------------------------------------
Фільм A (Екшн)       | 84.00              | 9.2736         | 0.9393
Фільм B (Комедія)    | 66.00              | 12.0830        | 0.5664
Фільм C (Драма)      | 90.00              | 9.4340         | 0.9892
----------------------------------------------------------------------
Красивий Вивід:


## Аналітичний розв'язок

**Дано вектори:**
* Користувач $\mathbf{u} = (8, 2, 5)$
* Фільм A $\mathbf{v}_A = (9, 1, 2)$
* Фільм B $\mathbf{v}_B = (1, 9, 8)$
* Фільм C $\mathbf{v}_C = (7, 2, 6)$

**Формула косинусної подібності:**
$$\cos \theta = \frac{\mathbf{u} \cdot \mathbf{v}}{\|\mathbf{u}\| \cdot \|\mathbf{v}\|}$$

---

### Крок 1. Обчислення норми (довжини) вектора користувача:
$$\|\mathbf{u}\| = \sqrt{8^2 + 2^2 + 5^2} = \sqrt{64 + 4 + 25} = \sqrt{93} \approx \mathbf{9.6437}$$

---

### Крок 2. Обчислення показників для кожного фільму:

#### 1. Фільм A (Екшн)
* **Скалярний добуток:** $8\cdot9 + 2\cdot1 + 5\cdot2 = 72 + 2 + 10 = \mathbf{84}$
* **Норма вектора:** $\sqrt{9^2 + 1^2 + 2^2} = \sqrt{86} \approx \mathbf{9.2736}$
* **Схожість:** $\frac{84}{9.6437 \cdot 9.2736} \approx \mathbf{0.9392}$

#### 2. Фільм B (Комедія)
* **Скалярний добуток:** $8\cdot1 + 2\cdot9 + 5\cdot8 = 8 + 18 + 40 = \mathbf{66}$
* **Норма вектора:** $\sqrt{1^2 + 9^2 + 8^2} = \sqrt{146} \approx \mathbf{12.0830}$
* **Схожість:** $\frac{66}{9.6437 \cdot 12.0830} \approx \mathbf{0.5664}$

#### 3. Фільм C (Драма)
* **Скалярний добуток:** $8\cdot7 + 2\cdot2 + 5\cdot6 = 56 + 4 + 30 = \mathbf{90}$
* **Норма вектора:** $\sqrt{7^2 + 2^2 + 6^2} = \sqrt{89} \approx \mathbf{9.4340}$
* **Схожість:** $\frac{90}{9.6437 \cdot 9.4340} \approx \mathbf{0.9892}$

### Висновок:
Найбільше значення косинусної подібності має **Фільм C (0.9892)**. Рекомендовано до перегляду.